### ISSUE TITLE
Activation recomputation should remember quantization recipes

### ISSUE THREAD

**Issue by kwyss-nvidia** [NVIDIA/TransformerEngine#1568]
At the time backward is invoked by `distributed.py` in the context of `activation_recompute=True` the FP8GlobalStateManager context is not the same as during the first invocation of the activations. Restoring the original quantization recipe would help ensure that the activations are recomputed consistently.

**Comment by ksivaman** [comment]
Is there a particular change in `FP8GlobalStateManager` that's causing an inconsistency in your workflow? The recipe should remain the same in the forward as well as the recompute phase. The FP8 metadata that does change (e.g. scale/amax in Delayed Scaling) is saved and restored in the recompute phase to ensure equivalent numerics.

**Comment by kwyss-nvidia** [comment]
That is not working as expected for new recipes.

Activation recompute is skipped for per-tensor current scaling here: https://github.com/NVIDIA/TransformerEngine/blob/8a20d666e4d63ff133621987b2bfe06e59711062/tests/pytorch/test_numerics.py#L674


I observed the default delayed scaling recipe being used when looking into this call site:
https://github.com/NVIDIA/TransformerEngine/pull/1559/commits/365a4d99be7671c3f5f6c5487d8961a109b18da5

**Comment by zhongbozhu** [comment]
I think if we are talking about checkpointing & recomputing, then we should also save the recipe so that wherever you call the loss.backward, it should remember what recipe it was using. 

Let's put the entire recompute feature aside, just talk about normal forward backward pass: the loss.backward is still outside of the fp8_autocast context, and it works because we have saved the recipe from fwd to bwd, and more importantly, all the quantizers are being saved. 

Same logic here, we cannot take it for granted that loss.backward should be put inside a fp8 autocast context, we should figure out a way to modify FP8GlobalStateManager when we are loading from the checkpoint and modify the state manager accordingly.

**Comment by kwyss-nvidia** [comment]
@zhongbozhu how does this look to store recipe and fp8 enabled? https://github.com/NVIDIA/TransformerEngine/pull/1559/commits/78c07efa9e7d4d00fe21d72e8dc461e67043321e

**Comment by kwyss-nvidia** [comment]
@ksivaman I have published a small MR to resolve this isolated from my large change: https://github.com/NVIDIA/TransformerEngine/pull/1577
